### PR TITLE
CreateDatasource

### PR DESCRIPTION
**Pull request by Tamarael**
Script to create data sources based off the "EncryptGatewayCredentials" PowerShell script

### CODE REVIEW

File: PowerShell Scripts/CreateDatasource.ps1 [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+$CreateDatasourceUrl = "https://api.powerbi.com/v1.0/myorg/gateways/$GatewayId/datasources"
+
+# Service Principal credentials
+$Secret = ""
+$AppID = "" 
+$AppPassword = ConvertTo-SecureString $Secret -AsPlainText -Force
+$Credential = New-Object PSCredential $AppID, $AppPassword
+$TenantID = ""
+
+# Sign in with Service Principal credentials
```
Comment (VsabGit) [inline PowerShell Scripts/CreateDatasource.ps1:27]: Acquire token for Service Principal

File: PowerShell Scripts/CreateDatasource.ps1 [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+Connect-PowerBIServiceAccount -Tenant $TenantID -ServicePrincipal -Credential $Credential
+
+# username = SQL user in case of basic credentials, or windows user in case of windows credentials
```
Comment (VsabGit) [inline PowerShell Scripts/CreateDatasource.ps1:30]: We need to specify that in this case we are giving a specific example for creating a datasource for an On-Prem SQL database using windows credentials.

Also we need to specify that the username here is the windows username and this is why it needs to contain the domain.

File: PowerShell Scripts/CreateDatasource.ps1 [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+# for SQL datasources, please be mindfull you will need to add the domain as in DOMAIN\user
+$username = ""
+# password = SQL user password in case of basic credentials, or windows user password in case of windows credentials
```
Comment (VsabGit) [inline PowerShell Scripts/CreateDatasource.ps1:33]: Windows password for the windows user. Here we need to emphasize that this needs to be a service account with minimal needed permissions on the database, we also want emphasize, that this is an example, and the password is written here as a plain text for convenience, in production we recommend that you use a KV and store the password there.

File: PowerShell Scripts/CreateDatasource.ps1 [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+$password = ''
+# Get the Gateway data for the encryption process
+$gw = Invoke-PowerBIRestMethod `
+                -Url "https://api.powerbi.com/v1.0/myorg/gateways/$GatewayId" `
+                -Method GET ` | ConvertFrom-Json                
+# On-Prem Gateway exponent which you can get using https://learn.microsoft.com/en-us/rest/api/power-bi/gateways/get-gateway API
```
Comment (VsabGit) [inline PowerShell Scripts/CreateDatasource.ps1:39]: "On-Prem Gateway exponent" no need for the rest of the sentence, you already implemented it.

File: PowerShell Scripts/CreateDatasource.ps1 [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+$gatewayExponent = $gw.publicKey.exponent
+
+# On-Prem Gateway modulus which you can get using https://learn.microsoft.com/en-us/rest/api/power-bi/gateways/get-gateway API
```
Comment (VsabGit) [inline PowerShell Scripts/CreateDatasource.ps1:42]: "On-Prem Gateway exponent" no need for the rest of the sentence, you already implemented it.

File: PowerShell Scripts/CreateDatasource.ps1 (hunk 1 above)
Comment (Tamarael) [reply PowerShell Scripts/CreateDatasource.ps1:27]: Done

File: PowerShell Scripts/CreateDatasource.ps1 (hunk 2 above)
Comment (Tamarael) [reply PowerShell Scripts/CreateDatasource.ps1:30]: done

File: PowerShell Scripts/CreateDatasource.ps1 (hunk 3 above)
Comment (Tamarael) [reply PowerShell Scripts/CreateDatasource.ps1:33]: done

**Review by Tamarael — COMMENTED** [review]
Changed the file according to the comments

File: PowerShell Scripts/CreateDatasource.ps1 (hunk 4 above)
Comment (Tamarael) [reply PowerShell Scripts/CreateDatasource.ps1:39]: done

Comment (Tamarael) [reply PowerShell Scripts/CreateDatasource.ps1:39]: done

**Review by VsabGit — APPROVED** [review]

**Pull request merged**